pilot/pkg/networking/core/v1alpha3/route/route_test.go

```diff
@@ -150,10 +151,24 @@ func TestBuildHTTPRoutes(t *testing.T) {
 		routes, err := route.BuildHTTPRoutesForVirtualService(node(cg), virtualServiceWithCatchAllPort,
 			serviceRegistry, nil, 8080, gatewayNames, route.RouteOptions{})
 		xdstest.ValidateRoutes(t, routes)
-
 		g.Expect(err).NotTo(gomega.HaveOccurred())
-		g.Expect(len(routes)).To(gomega.Equal(1))
+		g.Expect(len(routes)).To(gomega.Equal(3))
 		g.Expect(routes[0].Name).To(gomega.Equal("route 1.catch-all for 8080"))
+		g.Expect(routes[1].Name).To(gomega.Equal("route 2.header match"))
+		g.Expect(routes[2].Name).To(gomega.Equal("route 3"))
```

Comment:
Sorry I do not understand. If route 1 is "catch all", why do we need the others? They can never match?

Reply:
@howardjohn The reason why even the catchall route in the first it still works is here https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/route.go#L1450-L1463

Also verified below virtualservice spec it works in 1.10:
```
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: test
  namespace: test-pp
spec:
  gateways:
  - with-l7-rules-gw
  hosts:
  - '*'
  http:
  - match:
    - port: 80
      uri:
        prefix: /
    name: default-svc-80
    route:
    - destination:
        host: default-svc
        port:
          number: 8080
  - headers:
      request:
        add:
          X-CLIENT-IP: '%DOWNSTREAM_REMOTE_ADDRESS_WITHOUT_PORT%'
    match:
    - port: 80
      uri:
        prefix: /payment
    name: payment-80
    route:
    - destination:
        host: payment-svc
        port:
          number: 80
```
Corresponding envoy configdump:
```
     "version_info": "2023-06-01T02:31:24Z/126",
     "route_config": {
      "@type": "type.googleapis.com/envoy.config.route.v3.RouteConfiguration",
      "name": "https.443.default route",
      "virtual_hosts": [
       {
        "name": "*:443",
        "domains": [
         "*"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/checkout",
           "case_sensitive": true
          }
          ...
           },
           "max_grpc_timeout": "0s"
          },
          "name": "payment-svc-443-2"
         },
         {
          "match": {
           "prefix": "/",
           "case_sensitive": true
          },
         ...
            "host_selection_retry_max_attempts": "5",
            "retriable_status_codes": [
             503
            ]
           },
           "max_grpc_timeout": "0s"
          },

          "name": "default-svc-443-2"
         }
        ],
```